### PR TITLE
suricata: --include extra configuration files on the command line - v1

### PR DESCRIPTION
**Pull request by jasonish**
https://redmine.openinfosecfoundation.org/issues/3912

### CODE REVIEW

**Comment by suricata-qa** [comment]
WARNING:
 ERROR: ASAN TEST FAIL in build_asan QA test
 | field | baseline | test | % |
|--|--|--|--| 
| | build_asan | | |
 

Pipeline 10054

**Comment by jasonish** [comment]
Hmm. Some interesting some CI output.  Some obviously the result of the changes, others I don't see how.

File: src/suricata.c [hunk 1]
```diff
@@ -1679,6 +1688,21 @@ static TmEcode ParseCommandLine(int argc, char** argv, SCInstance *suri)
...
                 }
+            } else if (strcmp((long_opts[option_index]).name, "include") == 0) {
+                if (suri->additional_configs == NULL) {
+                    suri->additional_configs = SCCalloc(2, sizeof(char **));
+                    suri->additional_configs[0] = optarg;
+                } else {
+                    for (int i = 0;; i++) {
+                        if (suri->additional_configs[i] == NULL) {
+                            suri->additional_configs =
+                                    SCRealloc(suri->additional_configs, (i + 2) * sizeof(char **));
```
Comment (jasonish) [inline src/suricata.c:1699]: Will add error checking, but this is fatal in this stage right? As with the sccalloc above..

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 10056

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW1_stats_chk | | |
| .tcp.rst | 128100 | 104101 | 81.27% | 

Pipeline 10057

File: doc/userguide/partials/options.rst [hunk 2]
```diff
@@ -12,6 +12,14 @@
 
    Path to configuration file.
 
+.. option:: --include <path>
+
+   Additional configuration files to include. Multiple additonal
+   configuration files can be provided and will be included in the
+   order specified on the command line.  These additonal configuration
```
Comment (jufajardini) [inline doc/userguide/partials/options.rst:19]: nit: remove extra space before 'These'

Comment (jufajardini) [reply doc/userguide/partials/options.rst:19]: Can we add an example line showing how it would be done if one wanted to add multiple configuration files?

**Review by jufajardini — COMMENTED** [review]
Two suggestions, one I think may help clarify doc for users.

**Comment by jasonish** [comment]
> I like the functionality this provides. I would like to know if there's a way to provide a YAML fragment like the following to include the payload in the alert without having to fully specify the `fast` section and the entire `eve-log` section. IOW, can a single value be overridden without specifying the full YAML section?
> 
> ```
> %YAML 1.1
> ---                                                                                                                                                                                                                                                                                                        # Configure the type of alert (and other) logging you would like.                                                                                                                                                                                                                                          outputs:
>   # a line based alerts log similar to Snort's fast.log                                                                                                                                                                                                                                                      - fast:
>   # Extensible Event Format (nicknamed EVE) event log in JSON format                                                                                                                                                                                                                                         - eve-log:                                                                                                                                                                                                                                                                                                     
>       types:
>         - alert:                                                                                                                                                                                                                                                                                                       
>               payload: yes             # enable dumping payload in Base64
> ```

This is unrelated to overriding single values.  For single values, I was thinking of allowing a syntax where a fully qualified path would be provided in the YAML file:
```
this.that.parent.child: some-value
```
where the configuration loaded would descend into the current configuration and set that value.  The usual challenge for arrays applies, and I'm not sure how the payload example would be handled:
```
outputs.1.eve-log.types.0.alert.payload: yes
```
obviously this is fragile to edits on the configuration file, but I don't think there is a way around that other than reducing the usage of lists.

**Comment by victorjulien** [comment]
@jasonish can you rebase and resubmit? Would love this to go into 7.

**Comment by jasonish** [comment]
Rebased, replaced with https://github.com/OISF/suricata/pull/8438.

**Comment by jasonish** [comment]
> > I like the functionality this provides. I would like to know if there's a way to provide a YAML fragment like the following to include the payload in the alert without having to fully specify the `fast` section and the entire `eve-log` section. IOW, can a single value be overridden without specifying the full YAML section?
> > ```
> > %YAML 1.1
> > ---                                                                                                                                                                                                                                                                                                        # Configure the type of alert (and other) logging you would like.                                                                                                                                                                                                                                          outputs:
> >   # a line based alerts log similar to Snort's fast.log                                                                                                                                                                                                                                                      - fast:
> >   # Extensible Event Format (nicknamed EVE) event log in JSON format                                                                                                                                                                                                                                         - eve-log:                                                                                                                                                                                                                                                                                                     
> >       types:
> >         - alert:                                                                                                                                                                                                                                                                                                       
> >               payload: yes             # enable dumping payload in Base64
> > ```
> 
> This is unrelated to overriding single values. For single values, I was thinking of allowing a syntax where a fully qualified path would be provided in the YAML file:
> 
> ```
> this.that.parent.child: some-value
> ```
> 
> where the configuration loaded would descend into the current configuration and set that value. The usual challenge for arrays applies, and I'm not sure how the payload example would be handled:
> 
> ```
> outputs.1.eve-log.types.0.alert.payload: yes
> ```
> 
> obviously this is fragile to edits on the configuration file, but I don't think there is a way around that other than reducing the usage of lists.

https://redmine.openinfosecfoundation.org/issues/4783

**Pull request closed without merging**